### PR TITLE
Laboratory work #3, Vsevolod Papin - 23FPL2

### PR DESCRIPTION
**Pull request by AllKnowingGuy**
Thank you for closing the previous pull request!

### CODE REVIEW

File: lab_3_ann_retriever/settings.json [hunk 1]
```diff
@@ -1,3 +1,3 @@
 {
-    "target_score": 0
+    "target_score": 4
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/settings.json:2]: Don't forget to change the score

File: lab_3_ann_retriever/main.py [hunk 2]
```diff
@@ -3,10 +3,16 @@
 
 Vector search with text retrieving
 """
-
 # pylint: disable=too-few-public-methods, too-many-arguments, duplicate-code, unused-argument
+from json import dump, load
 from typing import Protocol
 
+from Scripts.rst2odt import output
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:10]: rm

File: lab_3_ann_retriever/main.py [hunk 3]
```diff
@@ -253,6 +388,18 @@ def index_documents(self, documents: list[str]) -> bool:
...
         In case of corrupt input arguments, False is returned.
         """
+        if not isinstance(documents, list):
+            return False
+
+        self._documents = documents
+        temp_vector_docs = []
+        for doc in self._documents:
+            vector_doc = self._index_document(doc)
+            if not vector_doc is Vector:
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:398]: can you use `isinstance()`?

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Great work! Increase your score

File: lab_3_ann_retriever/main.py (hunk 2 above)
Comment (AllKnowingGuy) [reply lab_3_ann_retriever/main.py:10]: Donr

File: lab_3_ann_retriever/main.py (hunk 3 above)
Comment (AllKnowingGuy) [reply lab_3_ann_retriever/main.py:398]: image
No, this causes an error

File: lab_3_ann_retriever/main.py [hunk 4]
```diff
@@ -426,6 +648,73 @@ def build(self, vectors: list[Vector]) -> bool:
...
+            if len(current_space["vectors"]) == 0:
+                info_list.pop(0)
+                continue
+
+            axis = current_space["depth"] % dimensions
+            current_space["vectors"].sort(key=lambda a: a[1][axis])
+            median_index = len(current_space["vectors"]) // 2
+            median_ind_vec = current_space["vectors"][median_index]
+            node_to_assign = Node(median_ind_vec[1], median_ind_vec[0])
+            """print("Axis and median:", axis, median_index)
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:675]: rm

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Great work! Fix comments and get ready to defend laboratory work 3

File: lab_3_ann_retriever/main.py [hunk 5]
```diff
@@ -49,6 +52,13 @@ def calculate_distance(query_vector: Vector, document_vector: Vector) -> float |
 
     In case of corrupt input arguments, None is returned.
     """
+    if not (isinstance(query_vector, tuple) and isinstance(document_vector, tuple)):
+        return None
+    if not (len(query_vector) > 0 and len(query_vector) == len(document_vector)):
+        return 0.0
+    size = len(document_vector)
+    sq_sum = sum((query_vector[ind] - document_vector[ind]) ** 2 for ind in range(size))
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:60]: do not create the variable

File: lab_3_ann_retriever/main.py [hunk 6]
```diff
@@ -61,6 +71,9 @@ def save_vector(vector: Vector) -> dict:
     Returns:
         dict: A state of the vector to save
     """
+    no_null_vec = {ind: value for ind, value in enumerate(vector) if value != 0.0}
+    vec_dict = {"len": len(vector),"elements": no_null_vec}
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:75]: the same

File: lab_3_ann_retriever/main.py [hunk 7]
```diff
@@ -104,6 +123,16 @@ def tokenize(self, text: str) -> list[str] | None:
...
         In case of corrupt input arguments, None is returned.
         """
+        if not isinstance(text, str):
+            return None
+        for symbol in text:
+            if symbol.isalpha() or symbol == " " or symbol == "\n":
+                continue
+            text = text.replace(symbol, " ")
+        no_stops_text = self._remove_stop_words(text.lower().split())
+        if isinstance(no_stops_text, list):
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:133]: reverse and return `None`

File: lab_3_ann_retriever/main.py [hunk 8]
```diff
@@ -130,13 +168,16 @@ def _remove_stop_words(self, tokens: list[str]) -> list[str] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not (isinstance(tokens, list) and len(tokens) > 0):
+            return None
+        new_tokens = [token for token in tokens if not token in self._stop_words]
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:173]: the same

File: lab_3_ann_retriever/main.py [hunk 9]
```diff
@@ -194,6 +276,16 @@ def save(self, file_path: str) -> bool:
...
         """
+        if not (isinstance(file_path, str) and len(file_path) > 0 and
+                isinstance(self._vocabulary, list) and isinstance(self._idf_values, dict) and
+                isinstance(self._token2ind, dict)):
+            return False
+        with open(file_path, "w", encoding="utf-8") as write_file:
+            to_write = {"vocabulary": self._vocabulary,
+                        "idf_values": self._idf_values,
+                        "token2ind": self._token2ind}
+            dump(to_write, write_file, indent="\t")
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:287]: better to import `json`

File: lab_3_ann_retriever/main.py [hunk 10]
```diff
@@ -207,6 +299,22 @@ def load(self, file_path: str) -> bool:
...
+            return False
+        with open(file_path, "r", encoding="utf-8") as read_file:
+            objects = load(read_file)
+        if not ("vocabulary" in objects and "idf_values" in objects and "token2ind" in objects):
+            return False
+        vocab = objects["vocabulary"]
+        idfs = objects["idf_values"]
+        token2ind = objects["token2ind"]
+        print(vocab, idfs, token2ind)
+        if isinstance(vocab, list) and isinstance(idfs, dict) and isinstance(token2ind, dict):
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:312]: reverse!

File: lab_3_ann_retriever/main.py [hunk 11]
```diff
@@ -304,6 +473,15 @@ def retrieve_vectorized(self, query_vector: Vector) -> str | None:
...
         In case of corrupt input arguments, None is returned.
         """
+        if not (isinstance(query_vector, tuple) and
+                len(query_vector) == len(self._document_vectors[0])):
+            return None
+        closest = self._calculate_knn(query_vector, self._document_vectors, 1)
+        if not (isinstance(closest, list) and len(closest) > 0):
+            return None
+        doc_index = closest[0][0]
+        vec_doc = self._documents[doc_index]
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:483]: do not create the variable

File: lab_3_ann_retriever/main.py [hunk 12]
```diff
@@ -389,6 +608,18 @@ def save(self) -> dict:
...
+        vec_to_save = save_vector(self.vector)
+        if not self.left_node is None:
+            left_node_saved = self.left_node.save()
+        else:
+            left_node_saved = None
+        if not self.right_node is None:
+            right_node_saved = self.right_node.save()
+        else:
+            right_node_saved = None
+        save_dict = {"vector": vec_to_save, "payload": self.payload,
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:620]: the same

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_ann_retriever/main.py (hunk 5 above)
Comment (AllKnowingGuy) [reply lab_3_ann_retriever/main.py]: Done

File: lab_3_ann_retriever/main.py (hunk 6 above)
Comment (AllKnowingGuy) [reply lab_3_ann_retriever/main.py]: Done

File: lab_3_ann_retriever/main.py (hunk 7 above)
Comment (AllKnowingGuy) [reply lab_3_ann_retriever/main.py:131]: Done

File: lab_3_ann_retriever/main.py (hunk 9 above)
Comment (AllKnowingGuy) [reply lab_3_ann_retriever/main.py]: Done

File: lab_3_ann_retriever/main.py (hunk 10 above)
Comment (AllKnowingGuy) [reply lab_3_ann_retriever/main.py:302]: Done

File: lab_3_ann_retriever/main.py (hunk 11 above)
Comment (AllKnowingGuy) [reply lab_3_ann_retriever/main.py]: Done

File: lab_3_ann_retriever/main.py (hunk 12 above)
Comment (AllKnowingGuy) [reply lab_3_ann_retriever/main.py]: Done

**Comment by AllKnowingGuy** [comment]
I really appreciate the idea with secrets, but I would be grateful if they were just a fun extra task without the bonus for the quickest reveal. This turned into a race that ended in a few days, and those who had troubles with pulling Lab 3 didn't have a chance for extra points.

**Pull request closed without merging**